### PR TITLE
Targeted recognition of trivial array operations

### PR DESCRIPTION
**Pull request by VermaSh**
Goal is to disallow induction variable elimination for trivial array operations as having those induction variables visible will probably lead to better optimizations later on.

We disallow the induction variable reduction if:
- taken side of the if is low frequency
- load and store to local variable
- tree top which anchors stuff
- NULLCHK, BNDCHK and array store check
- aladd under loadi or storei which has mathematical expressions based on induction variables and constants

### CODE REVIEW

**Comment by VermaSh** [comment]
I am looking into some failures relating to node visit count. I'll post another update soon. I'll clean up the code (create a list of allowed nodes, move the method to it's own class) once I have the prototype working

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (fjeremic) [inline compiler/optimizer/LoopCanonicalizer.cpp:2150]: This seems too restrictive. What about trivial array operations on non-byte arrays? Those will not be ignored by this `if`. We should instead use the opcode APIs, for example `isStoreIndirect`, `isConversion`, `isAdd`, etc.

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (fjeremic) [inline compiler/optimizer/LoopCanonicalizer.cpp:2140]: This looks like a logic typo. Did you mean to use `==` here instead of `=`? I'm surprised this even compiles.

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (VermaSh) [reply compiler/optimizer/LoopCanonicalizer.cpp:2150]: Right, that's because I wanted to get the recognition working for the set of trees we are seeing in the benchmark first. But I'll definitely replace the opcode with APIs

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (VermaSh) [reply compiler/optimizer/LoopCanonicalizer.cpp:2140]: Ah dang, that was definitely not me, it was vs code auto fill!

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (andrewcraik) [reply compiler/optimizer/LoopCanonicalizer.cpp:2150]: I think it would be better to use capability based queries so we can handle aladd and aidd, iadd and isub etc and make this less brittle across different modes. I know you want to keep it simple, but I agree this seems a bit overfit to the specific case you are looking at.

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (fjeremic) [inline compiler/optimizer/LoopCanonicalizer.cpp:2007]: This variable needs to be `NULL` initialized.

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (fjeremic) [inline compiler/optimizer/LoopCanonicalizer.cpp:2009]: This variable is not used outside the `for` loop. Can we move the definition there?

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (fjeremic) [inline compiler/optimizer/LoopCanonicalizer.cpp:2016]: Do we always expect to hit this path? Is there a possibility we never execute this `goto`?

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (fjeremic) [inline compiler/optimizer/LoopCanonicalizer.cpp:2026]: This should be guarded via `performTransformation` as it is a transformation after all (i.e. we are choosing not to do something we otherwise would).

File: compiler/optimizer/LoopCanonicalizer.hpp
Comment (fjeremic) [inline compiler/optimizer/LoopCanonicalizer.hpp:259]: Does this need to exist as a member function? It does not appear to use any instance fields of this class so it could be defined as a static function in the compilation unit.

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (fjeremic) [inline compiler/optimizer/LoopCanonicalizer.cpp:2133]: We should turn this into proper Doxygen documentation, how about the following:

```
/**
 * \brief
 *    Determines whether the node is a trivial (arithmetic and checks only) array operation.
 *
 * \param visited
 *    The list of visited descendants which should not be examined.
 *
 * \param node
 *    The node to examine.
 *
 * \return
 *    \c true if \p node is a trivial array operation; \c false otherwise.
 *
 * \details
 *    This API is used to identify trivial array operations while performing induction variable elimination. In cases
 *    where the array operation is considered trivial there may be a negative trade-off between eliminating an extra
 *    induction variable compared to leaving it present. The typical scenario this can be encountered in is an array
 *    copy operation which may look like this (in pseudocode):
 *
 *    ```
 *    while (j < x) {
 *        a[i++] = b[j++];
 *    }
 *    ```
 *
 *    If induction variables `i` and `j` are not known to be constants outside the loop then eliminating one induction
 *    variable forces us to rewrite the above loop to be:
 *
 *    ```
 *    temp = i - j;
 *    while (j < x) {
 *        a[temp + j] = b[j++]
 *    }
 *    i = temp + j;
 *    ```
 *
 *    i.e. we have to rewrite one induction variable in terms of the other. The latter loop is more complex than the
 *    former because the recalculation of the induction variable now dominates the computation within the loop and
 *    hence it is not a worth while trade-off. Performing such an induction variable elimination can result in degraded
 *    performance which could then be amplified by loop unrolling at later stages of the optimization strategy.
 */
bool TR_LoopCanonicalizer::isTrivialArrayOperation(TR::NodeChecklist &visited, TR::Node *node)
   {
```

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (VermaSh) [reply compiler/optimizer/LoopCanonicalizer.cpp:2016]: Only time we wouldn't want to execute `trivialArrayOperationCheck` is if we didn't find the node containing the derived induction variable. Don't think that is possible. So based on that, we do always hit that path

Comment (fjeremic) [reply compiler/optimizer/LoopCanonicalizer.cpp:2016]: In that case I'd like to see the assert on line 2021 removed and changed to:

```
TR_ASSERT_FATAL_WITH_NODE(node, node != NULL, "Unable to find node containing the derived induction variable");
```

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (andrewcraik) [inline compiler/optimizer/LoopCanonicalizer.cpp:2094]: I'm not a huge fan of this long-range goto - I'm not quite sure how to change the code structure to avoid it, but it is not very nice for reading and understanding the code.

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (andrewcraik) [inline compiler/optimizer/LoopCanonicalizer.cpp:1881]: given this check code will cause us to fail to do something I think this needs some more trace to explain which tree and why we gave up on a particular operation. Debugging from only a log is going to be pretty hard without some more trace IMO.

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (andrewcraik) [inline compiler/optimizer/LoopCanonicalizer.cpp:2034]: Unless you are updating this list I think you want to look at using the subnode iterator on the region https://github.com/eclipse/omr/blob/23ce7a13d1b1decbe61bea5ee08e08339b359524/compiler/optimizer/Structure.hpp#L611 to save copying the pointers into the list only to iterate them again.

Comment (andrewcraik) [reply compiler/optimizer/LoopCanonicalizer.cpp:2034]: When using the iterator you can say that you know your trivial loop check won't return true if you encounter a nested loop or improper region.

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (andrewcraik) [reply compiler/optimizer/LoopCanonicalizer.cpp:2083]: why not just add derivedInductionVarNode to the loop conditions so you don't need that ugly goto?

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (VermaSh) [reply compiler/optimizer/LoopCanonicalizer.cpp:2094]: I could use the result from `isTrivialArrayOperation(..)` to set `incrementsInLockStep` https://github.com/eclipse/omr/blob/f4ab45dc92a334b143d5c434f441a3c950e935a6/compiler/optimizer/LoopCanonicalizer.cpp#L2109
if it is a trivial array operation `incrementsInLockStep` would be set to false and we would skip the next two if blocks

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (VermaSh) [reply compiler/optimizer/LoopCanonicalizer.cpp]: I have added some tracing in the latest commit. Let me know if you would like to see more tracing

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (fjeremic) [inline compiler/optimizer/LoopCanonicalizer.cpp:2105]: Check other uses of `performTransformation`. This API is used to guard transformations which can then be searched via `lastOptTransformationIndex` or `lastOptSubindex`. In your current implementation we will proceed with the transformation irregardless because no code is being guarded by the return value of `performTransformation`. We need to change this to something like:

```
if (istrivialArrayOperation)
   istrivialArrayOperation = performTransformation(comp(), "%sSkipping...
```

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (fjeremic) [inline compiler/optimizer/LoopCanonicalizer.cpp:2103]: Spaces after arguments are missing here and many other places in this PR.

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (VermaSh) [reply compiler/optimizer/LoopCanonicalizer.cpp:2105]: ah I see your point. But if we disable `trivialArrayOperationCheck` opt using `lastOptTransformationIndex` or `lastOptSubindex`. That would also disable the redundant var elimination transformation here https://github.com/eclipse/omr/blob/53ce7df9200f1e308c8ad2d5a12dd8795b61fd14/compiler/optimizer/LoopCanonicalizer.cpp#L2148  since they are mutually exclusive.

I think we should look at the `trivialArrayOperationCheck` not as a transformation but rather as a input/criteria used to evaluate cost/benefit of redundant var elimination.

**Comment by VermaSh** [comment]
There is a lot of commented code in that function. Should I just delete that and submit it in this PR?

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (andrewcraik) [inline compiler/optimizer/LoopCanonicalizer.cpp:1894]: This needs to be guarded so maybe make this an instance method so you can guard with `if (trace())`

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (andrewcraik) [inline compiler/optimizer/LoopCanonicalizer.cpp:1896]: You probably want to check the opcode is operating on an integer, long or address and not a floating point or some other type...

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (andrewcraik) [inline compiler/optimizer/LoopCanonicalizer.cpp:1907]: guard with an appropriate `if (trace())` or similar.

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (andrewcraik) [reply compiler/optimizer/LoopCanonicalizer.cpp:2105]: Skipping the transform needs to be exposed to the triage/binary search process somehow so you need to consider that and do something appropriate here, whatever that is.

Comment (VermaSh) [reply compiler/optimizer/LoopCanonicalizer.cpp:2105]: Thinking about it now, it's definitely better to disable both transformations rather than disable one only to find out other was enabled the entire time.

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (VermaSh) [reply compiler/optimizer/LoopCanonicalizer.cpp:1894]: Would it be appropriate to replace `traceMsg` with `dumpOptDetails`?

Comment (andrewcraik) [reply compiler/optimizer/LoopCanonicalizer.cpp:1894]: no - dumpOptDetails is the non-optional equivalent of performTransformation - eg performTransformation is used to guard a transform that is optional. Sometimes a transform is not optional - it might be necessary for correctness or it might be the follow-on to a previous optional change - in these cases we use dumpOptDetails to record the non-optional transform. This is just a trace statement and so does not need the notion of a transformation attached to it

**Comment by VermaSh** [comment]
There are two test failures on macOS, but don't seem to be caused by these changes
```
31: [  PASSED  ] 233 tests.
31: [  FAILED  ] 2 tests, listed below:
31: [  FAILED  ] PortSockTest.poll_functionality_basic
31: [  FAILED  ] PortSockTest.poll_functionality_many_sockets
```

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (fjeremic) [inline compiler/optimizer/LoopCanonicalizer.cpp:2081]: This comment seems unnecessary.

**Review by fjeremic — APPROVED** [review]

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (VermaSh) [reply compiler/optimizer/LoopCanonicalizer.cpp]: I have updated the commit

**Review by fjeremic — APPROVED** [review]

**Comment by VermaSh** [comment]
I've addressed all outstanding reviews, please let me know if there are any other changes that should be made.

**Comment by andrewcraik** [comment]
Code looks good - thanks for the revisions! Before I approve could you comment on the perf testing that has been done now that the code has been updated?

**Comment by VermaSh** [comment]
@andrewcraik With these changes the throughput is back on par with last known good build

**Comment by andrewcraik** [comment]
have you done any verification beyond just the regression to make sure you aren't causing more general regressions with the exclusion?

**Comment by fjeremic** [comment]
> have you done any verification beyond just the regression to make sure you aren't causing more general regressions with the exclusion?

@VermaSh have we verified the current state of this PR fixes the performance problem you were observing?

**Comment by VermaSh** [comment]
@fjeremic Yes, current changes fix the performance problems we were facing previously

**Comment by fjeremic** [comment]
Preemptively launching builds here given both committers are satisfied.

@genie-omr build all

**Comment by fjeremic** [comment]
@andrewcraik this one should be ready.

**Comment by fjeremic** [comment]
@andrewcraik this one should be ready.

**Comment by andrewcraik** [comment]
I have followed up with @VermaSh to clarify my concern was that performance of benchmarks other than the one that had the problem had been verified to prevent accidental collateral damage from the change. I believe this is being followed up so I am going to hold merging until there is confirmation that there are no known side-effects in terms of performance from the change.

**Comment by fjeremic** [comment]
@VermaSh any update on this one?

**Comment by VermaSh** [comment]
The changes seem to be causing regression in other benchmarks. I am currently in the process of verifying the regression on dedicated machines.

File: compiler/optimizer/LoopCanonicalizer.cpp
Comment (andrewcraik) [inline compiler/optimizer/LoopCanonicalizer.cpp:2104]: This trace needs to include a loop number or something as well to make it easier to follow the log without the opt being traced.

Comment (VermaSh) [reply compiler/optimizer/LoopCanonicalizer.cpp:2111]: Apologies for the delay. The loop number is already being printed to the trace log here https://github.com/eclipse/omr/blob/4d32dfda3c999f4e1c27b7dcca73ab27c1989a0f/compiler/optimizer/LoopCanonicalizer.cpp#L1956

**Comment by VermaSh** [comment]
Closing this as we have deiced to use a different approach to fix the regression

**Pull request closed without merging**
